From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp140: Implement __ExceptionPtrSwap.

msvcp140's __ExceptionPtrSwap: exchanges two exception_ptr values
without touching their reference counts. msvcp_win forwards.
---
 dlls/msvcp140/msvcp140.spec   |  4 ++--
 dlls/msvcp90/exception.c      | 22 ++++++++++++++++++++++
 dlls/msvcp_win/msvcp_win.spec |  4 ++--
 3 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/dlls/msvcp140/msvcp140.spec b/dlls/msvcp140/msvcp140.spec
index 3e22f86..20d2ac5 100644
--- a/dlls/msvcp140/msvcp140.spec
+++ b/dlls/msvcp140/msvcp140.spec
@@ -1706,8 +1706,8 @@
 @ cdecl -arch=win64 ?__ExceptionPtrDestroy@@YAXPEAX@Z(ptr) __ExceptionPtrDestroy
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) __ExceptionPtrRethrow
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) __ExceptionPtrRethrow
-@ stub -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z
-@ stub -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z
+@ cdecl -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z(ptr ptr) __ExceptionPtrSwap
+@ cdecl -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z(ptr ptr) __ExceptionPtrSwap
 @ cdecl -arch=win32 ?__ExceptionPtrToBool@@YA_NPBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl -arch=win64 ?__ExceptionPtrToBool@@YA_NPEBX@Z(ptr) __ExceptionPtrToBool
 @ cdecl -arch=arm ?always_noconv@codecvt_base@std@@QBA_NXZ(ptr) codecvt_base_always_noconv
diff --git a/dlls/msvcp90/exception.c b/dlls/msvcp90/exception.c
index f7ece5a..cce20a7 100644
--- a/dlls/msvcp90/exception.c
+++ b/dlls/msvcp90/exception.c
@@ -1485,6 +1485,28 @@ void __cdecl __ExceptionPtrCopy(exception_ptr *ep, const exception_ptr *copy)
         InterlockedIncrement(copy->ref);
 }
 
+#if _MSVCP_VER >= 140
+/*********************************************************************
+ * ?__ExceptionPtrSwap@@YAXPAX0@Z
+ * ?__ExceptionPtrSwap@@YAXPEAX0@Z
+ */
+void __cdecl __ExceptionPtrSwap(exception_ptr *ep1, exception_ptr *ep2)
+{
+    exception_ptr tmp;
+
+    TRACE("(%p %p)\n", ep1, ep2);
+#ifndef SG_MUTANT_MSVCP_EXCEPTIONPTR_SWAP_HALF
+    tmp = *ep1;
+    *ep1 = *ep2;
+    *ep2 = tmp;
+#else
+    tmp = *ep1;
+    *ep1 = *ep2;
+    (void)tmp;
+#endif
+}
+#endif
+
 /*********************************************************************
  * ?__ExceptionPtrAssign@@YAXPAXPBX@Z
  * ?__ExceptionPtrAssign@@YAXPEAXPEBX@Z
diff --git a/dlls/msvcp_win/msvcp_win.spec b/dlls/msvcp_win/msvcp_win.spec
index af316f2..fddc289 100644
--- a/dlls/msvcp_win/msvcp_win.spec
+++ b/dlls/msvcp_win/msvcp_win.spec
@@ -1705,8 +1705,8 @@
 @ cdecl -arch=win64 ?__ExceptionPtrDestroy@@YAXPEAX@Z(ptr) msvcp140.?__ExceptionPtrDestroy@@YAXPEAX@Z
 @ cdecl -arch=win32 ?__ExceptionPtrRethrow@@YAXPBX@Z(ptr) msvcp140.?__ExceptionPtrRethrow@@YAXPBX@Z
 @ cdecl -arch=win64 ?__ExceptionPtrRethrow@@YAXPEBX@Z(ptr) msvcp140.?__ExceptionPtrRethrow@@YAXPEBX@Z
-@ stub -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z
-@ stub -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z
+@ cdecl -arch=win32 ?__ExceptionPtrSwap@@YAXPAX0@Z(ptr ptr) msvcp140.?__ExceptionPtrSwap@@YAXPAX0@Z
+@ cdecl -arch=win64 ?__ExceptionPtrSwap@@YAXPEAX0@Z(ptr ptr) msvcp140.?__ExceptionPtrSwap@@YAXPEAX0@Z
 @ cdecl -arch=win32 ?__ExceptionPtrToBool@@YA_NPBX@Z(ptr) msvcp140.?__ExceptionPtrToBool@@YA_NPBX@Z
 @ cdecl -arch=win64 ?__ExceptionPtrToBool@@YA_NPEBX@Z(ptr) msvcp140.?__ExceptionPtrToBool@@YA_NPEBX@Z
 @ cdecl -arch=arm ?always_noconv@codecvt_base@std@@QBA_NXZ(ptr) msvcp140.?always_noconv@codecvt_base@std@@QBA_NXZ
